pyro/infer/autoguide/guides.py

```diff
@@ -1482,3 +1506,346 @@ def median(self, *args, **kwargs):
+        init_param
+            Not implemented yet - initial values for amortised variables.
+        init_param_scale
+            How to scale/normalise initial values for weights converting hidden layers to mean and sd.
+        data_transform
+            Function to use for transforming data before passing it to encoder network.
+        encoder_class
+            Class for defining encoder network.
+        encoder_kwargs
+            Keyword arguments for encoder_class.
+        create_plates
+            Function for creating plates
+        single_encoder
+            Use single encoder for all variables (True) or one encoder per variable (False).
+        """
+
+        super().__init__(model, create_plates=create_plates)
+        self.amortised_plate_sites = amortised_plate_sites
+        self.single_encoder = single_encoder
+
+        self.softplus = SoftplusTransform()
+
+        encoder_kwargs = encoder_kwargs if isinstance(encoder_kwargs, dict) else dict()
+        encoder_kwargs["n_hidden"] = n_hidden
+        self.encoder_kwargs = encoder_kwargs
+
+        self.n_in = n_in
+        self.n_out = (
+            np.sum(
+                [
+                    np.sum(amortised_plate_sites["sites"][k])
+                    for k in amortised_plate_sites["sites"].keys()
+                ]
+            )
+            * 2
+        )
+        self.n_hidden = n_hidden
+        self.encoder_class = encoder_class
+        if self.single_encoder:
+            # create a single encoder NN
+            self.encoder = encoder_class(
+                n_in=self.n_in, n_out=self.n_hidden, **self.encoder_kwargs
+            )
+
+        self.init_param_scale = init_param_scale
+        self.data_transform = data_transform
+
+    def _setup_prototype(self, *args, **kwargs):
+
+        super()._setup_prototype(*args, **kwargs)
+
+        self._event_dims = {}
+        self._cond_indep_stacks = {}
+        self.hidden2locs = PyroModule()
+        self.hidden2scales = PyroModule()
+
+        if not self.single_encoder:
+            # create module class for collecting multiple encoder NN
+            self.encoder = PyroModule()
+
+        # Initialize guide params
+        for name, site in self.prototype_trace.iter_stochastic_nodes():
+            # Collect unconstrained event_dims, which may differ from constrained event_dims.
+            with helpful_support_errors(site):
+                init_loc = (
+                    biject_to(site["fn"].support).inv(site["value"].detach()).detach()
+                )
+            event_dim = site["fn"].event_dim + init_loc.dim() - site["value"].dim()
+            self._event_dims[name] = event_dim
+
+            # Collect independence contexts.
+            self._cond_indep_stacks[name] = site["cond_indep_stack"]
+
+            # add linear layer for locs and scales
+            param_dim = (self.n_hidden, self.amortised_plate_sites["sites"][name])
```

Comment:
One of the main assumptions at the moment is that encoded variables are 2D tensors of shape (plate subsample_size aka batch size, self.amortised_plate_sites["sites"][name]) - but I guess that the shape can be automatically guessed, I just did not think that through and do not have applications where variables are more than 2D.

Reply:
I think it's more important to get an initial version merged quickly than to make something fully general right away.  So WDYT about simply adding assertions or `NotImplementedError("not yet supported")` checks for your current assumptions?

Also feel free to start the class docstring with `EXPERIMENTAL and add a `.. warning:: Interface may change` to give you/us room to slightly change the interface later in case that fully-general version needs slight changes?